kube-client/src/api/portforward.rs

```diff
@@ -0,0 +1,335 @@
+            result: None,
+        }));
+        let shared_state = state.clone();
+        let port_nums = port_nums.to_owned();
+        tokio::spawn(async move {
+            let result = start_message_loop(stream, port_nums, duplexes, errors).await;
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.result = Some(result);
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
+            }
+        });
+        Portforwarder { ports, state }
+    }
+
+    /// Get streams for forwarded ports.
+    pub fn ports(&mut self) -> &mut [Port] {
+        self.ports.as_mut_slice()
+    }
+}
+
+impl Future for Portforwarder {
+    type Output = Result<(), Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut state = self.state.lock().unwrap();
+        if let Some(result) = state.result.take() {
+            return Poll::Ready(result);
+        }
+
+        if let Some(waker) = &state.waker {
+            if waker.will_wake(cx.waker()) {
+                return Poll::Pending;
+            }
+        }
+
+        state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+pub struct Port {
+    // Data pipe.
+    stream: Option<DuplexStream>,
+    // Error channel.
+    error: Option<ErrorReceiver>,
+}
+
+impl Port {
+    pub(crate) fn new(stream: DuplexStream, error: ErrorReceiver) -> Self {
+        Port {
+            stream: Some(stream),
+            error: Some(error),
+        }
+    }
+
+    /// Data pipe for sending to and receiving from the forwarded port.
+    ///
+    /// This returns a `Some` on the first call, then a `None` on every subsequent call
+    pub fn stream(&mut self) -> Option<impl AsyncRead + AsyncWrite + Unpin> {
+        self.stream.take()
+    }
+
+    /// Future that resolves with any error message or when the error sender is dropped.
+    /// When the future resolves, the port should be considered no longer usable.
+    ///
+    /// This returns a `Some` on the first call, then a `None` on every subsequent call
+    pub fn error(&mut self) -> Option<impl Future<Output = Option<String>>> {
+        // Ignore Cancellation error.
+        self.error.take().map(|recv| recv.map(|res| res.ok()))
+    }
+}
+
+async fn start_message_loop<S>(
+    stream: WebSocketStream<S>,
+    ports: Vec<u16>,
+    duplexes: Vec<DuplexStream>,
+    error_senders: Vec<Option<ErrorSender>>,
+) -> Result<(), Error>
+where
+    S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+{
+    let mut writers = Vec::new();
+    // Loops to run concurrently.
+    // We can spawn tasks to run `to_pod_loop` in parallel and flatten the errors, but the other 2 loops
+    // are over a single WebSocket connection and cannot process each port in parallel.
+    let mut loops = Vec::with_capacity(ports.len() + 2);
+    // Channel to communicate with the main loop
+    let (sender, receiver) = mpsc::channel::<Message>(1);
```

Comment:
AFAIKT this gives us one guaranteed send slot. Not sure if this leaves it possible for us to overwrites the last message before sending our end. I don't think so; we read one-by-one from both sides, and send one-by-one so hopefully tokio flushes each message out correctly - but not sure how these things truly propagate below.

Reply:
Maybe you're thinking of `tokio::sync::mpsc::channel`?

This is `futures::channel::mpsc::channel` and the [documentation](https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html) says:

> The channel’s capacity is equal to buffer + num-senders. In other words, each sender gets a guaranteed slot in the channel capacity, and on top of that there are buffer “first come, first serve” slots available to all senders.

Reply:
The channel is shared between ports though, so while each port is guaranteed to be able to have one slot in there.

..but looking into it, it does not look like it's a problem because each `.send` is awaited. For some reason I thought it was just dropped into a ring buffer inside the channel.